src/components/text-editor/prosemirror-adapter/plugins/trigger/inserter.ts

```diff
@@ -9,36 +9,59 @@ export const inserterFactory = (view: EditorView): TextEditor => {
     const startPos = getTriggerStartPosition(view);
 
     return {
-        insert: (input: TextEditorNode | string) => {
-            const { state, dispatch } = view;
-            const { schema, selection } = state;
-            const { $from } = selection;
+        insert: insert(view, startPos),
+        insertHtml: insertHtml(view, startPos),
+    };
+};
 
-            let node: Node;
+const insert =
+    (view: EditorView, startPos: number) =>
+    (input: TextEditorNode | string): void => {
+        const schema = view.state.schema;
+        let node: Node;
 
-            try {
-                node = createNode(input, schema);
-            } catch (error) {
-                // eslint-disable-next-line no-console
-                console.error(error.message);
+        try {
+            node = createNode(input, schema);
+        } catch (error) {
+            // eslint-disable-next-line no-console
+            console.error(error.message);
 
-                return;
-            }
+            return;
+        }
 
-            const spaceNode = schema.text(' ');
+        const spaceNode = schema.text(' ');
 
-            const fragment = schema.nodes.doc.create(null, [node, spaceNode]);
+        const fragment = schema.nodes.doc.create(null, [node, spaceNode]);
+
+        dispatchTransaction(view, startPos, fragment);
+    };
 
-            const transaction = state.tr.replaceWith(
-                startPos,
-                $from.pos,
-                fragment,
-            );
+const insertHtml =
+    (view: EditorView, startPos: number) =>
+    (input: string): void => {
+        const schema = view.state.schema;
 
-            transaction.setMeta('stopTrigger', true);
-            dispatch(transaction);
-        },
+        const container = document.createElement('div');
+        container.innerHTML = input;
```

Comment:
This seems very dangerous and is a potential vector for XSS attacks. Is the input properly sanitized before we use it here?

Reply:
Good point.

For our use case it will be, because the html string has always been through the text-editor once.
For other cases, I guess that can't be guaranteed :(

I assume this might be a blocker. Should we sanitize the input? Do you have other ideas to solve the root issue?

Reply:
In our parser we use the markdownToHTML pipeline from the markdown component. It contains a sanitizer. The input could be run through that before being inserted here. 

https://github.com/Lundalogik/lime-elements/blob/ed28b485d74d9199df1f51dcf2a37eb8d59f6d85/src/components/markdown/markdown-parser.ts#L29

Reply:
I put in a PR that bumped prosemirror-model to version 1.22.1 earlier this week after reading that same forum post. I've also tested injecting sneaky scripts in via both html, script tags and also in via css as a value on a css property. So far everything was cleaned and removed if it wasn't kosher. 

It seems we only run the html through the markdownToHTML when we parse between HTML and Markdown. 👀 
The `ContentTypeConverter` was one of the first things built into our `prosemirror-adapter`. I never really noticed that it was simply returning a Promise with the string. 

We could reuse the markdownToHTML pipeline as we do here in the markdown parser:
https://github.com/Lundalogik/lime-elements/blob/ed28b485d74d9199df1f51dcf2a37eb8d59f6d85/src/components/text-editor/utils/markdown-converter.ts#L72

Reply:
I pushed a fixup to run the input through the `markdownToHTML` pipeline.

Not entirely sure I follow different converters/parsers that are involved here. Maybe you guys can refactor this as a next step if necessary? 👼